### PR TITLE
7 post /api/articles/id/comments

### PR DESCRIPTION
**Pull request by tomalcock**

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -141,3 +141,51 @@ describe("GET /api/articles", () => {
...
+            body: "this is a test"
+        }
+        return request(app)
+        .post('/api/articles/3/comments')
+        .send(newComment)
+        .expect(201)
+        .then((response) => {
+            const newComm = response.body.comment
+            console.log(newComm)
+            expect(newComm.author).toBe('butter_bridge');
```
Comment (Harrison365) [inline __tests__/app.test.js:156]: Good work :) 
Please remove console.logs when you're done with them 👍

File: models/comments.models.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+const db = require('../db/connection.js')
+const format = require('pg-format');
+
+function insertComment({username,body},article_id) {
+    return db
+    .query(
+    `INSERT INTO comments (body, article_id, author) VALUES ($1,$2,$3) RETURNING author,body;`,
+    [body,article_id,username]
+    )
```
Comment (Harrison365) [inline models/comments.models.js:9]: You can do a return * here

File: __tests__/app.test.js [hunk 3]
```diff
@@ -141,3 +141,51 @@ describe("GET /api/articles", () => {
...
+            expect(newComm.body).toBe('this is a test');
+        })
+    })
```
Comment (Harrison365) [inline __tests__/app.test.js:162]: Please test all comment properties e.g. comment_id etc. (see comment on SQL in model 👍

File: __tests__/app.test.js [hunk 4]
```diff
@@ -141,3 +141,51 @@ describe("GET /api/articles", () => {
...
+        const newComment = {
+            user: 'butter_bridge',
+            bodyyyy: "this is a test"
+        }
+        return request(app)
+          .post('/api/articles/3/comments')
+          .send(newComment)
+          .expect(400)
+          .then((response) => {
+            expect(response.body.msg).toBe('Bad Request');
```
Comment (Harrison365) [inline __tests__/app.test.js:188]: Nice 👍 

Please add the following tests:
- 400 invalid article id e.g. banana
- 404 username not found

File: __tests__/app.test.js [hunk 5]
```diff
@@ -141,3 +141,51 @@ describe("GET /api/articles", () => {
...
+            username: 'butter_bridge',
+            body: "this is a test"
+        }
+        return request(app)
+          .post('/api/articles/999/comments')
+          .send(newComment)
+          .expect(400)
+          .then((response) => {
+            expect(response.body.msg).toBe('Bad Request');
+          });
```
Comment (Harrison365) [inline __tests__/app.test.js:175]: This would be a 404 article valid but not found.

**Comment by Harrison365** [comment]
Nice work 👍 
Just a few more tests to add/alter 
Please see comments :)

File: __tests__/app.test.js [hunk 6]
```diff
@@ -141,3 +141,80 @@ describe("GET /api/articles", () => {
...
+        return request(app)
+        .post('/api/articles/3/comments')
+        .send(newComment)
+        .expect(201)
+        .then((response) => {
+            const newComm = response.body.comment
+            expect(newComm.author).toBe('butter_bridge');
+            expect(newComm.body).toBe('this is a test');
+            expect(newComm.article_id).toBe(3);
+            expect(newComm.votes).toBe(0);
```
Comment (Pipe-Smoking-Rabbit) [inline __tests__/app.test.js:159]: This is a good test, you could also add an assertion that you expect the property `created_at` to exist on the object

File: app.js [hunk 7]
```diff
@@ -24,14 +27,30 @@ app.get("/api/articles/:article_id", getArticleByID);
...
+    //console.log(err.code)
+    if(err.status && err.msg) {
+        res.status(err.status).send({msg: err.msg})
+    }
+    if(err.length === 240) {
         res.status(400).send({msg: 'Bad Request'})
     }
+    if(err.constraint === 'comments_article_id_fkey') {
+        res.status(404).send({msg: 'not found'})
+    }
```
Comment (Pipe-Smoking-Rabbit) [inline app.js]: Line `42` and `45` shouldn't be necessary, I wouldn't advise handling errors by the length... handling by the constraint could be situationally okay 
If you removed both these if statements I believe your tests should all still pass ?

File: app.js [hunk 8]
```diff
@@ -24,14 +27,30 @@ app.get("/api/articles/:article_id", getArticleByID);
...
+    if(err.code === '22P02'|| err.code === '23503') {
+        res.status(400).send({msg: 'Bad Request'})
+    }
+    if(err.code === '23502') {
```
Comment (Pipe-Smoking-Rabbit) [inline app.js]: Here's your issue, `23502` and `23503` need to be swapped around:
- `23502` - violation of a 'not null' constraint which is _400_
- `23503` - foreign key violation which is _404_

**Review by Pipe-Smoking-Rabbit — APPROVED** [review]
I've identified your bug in the error handling in the comments below 🐛  
It all looks good besides that, you could add one more test for:
- 200: endpoint ignores unnecessary keys and completes post provided it has username and body

Don't worry about resubmitting though , and don't forget to remove all the console logs before merging 🧹

**Pull request merged**